tpm2/policy.go

```diff
@@ -151,189 +152,263 @@ func makeStaticPolicyDataRaw_v1(data *staticPolicyData) *staticPolicyDataRaw_v1
-		}
-		signature = tpm2.Signature{
-			SigAlg: tpm2.SigSchemeAlgECDSA,
-			Signature: &tpm2.SignatureU{
-				ECDSA: &tpm2.SignatureECDSA{
-					Hash:       signDigest,
-					SignatureR: sigR.Bytes(),
-					SignatureS: sigS.Bytes()}}}
-	default:
-		panic("invalid private key type")
-	}
 
 	// Load the public part of the key in to the TPM. There's no integrity protection for this command as if it's altered in
 	// transit then either the signature verification fails or the policy digest will not match the one associated with the NV
 	// index.
-	keyLoaded, err := tpm.LoadExternal(nil, keyPublic, tpm2.HandleEndorsement)
+	keyLoaded, err := tpm.LoadExternal(nil, c.updateKey, tpm2.HandleEndorsement)
 	if err != nil {
 		return xerrors.Errorf("cannot load public part of key used to verify authorization signature: %w", err)
 	}
 	defer tpm.FlushContext(keyLoaded)
 
-	// Execute the policy assertions
-	if version == 0 {
-		// See the comment for computeV0PinNVIndexPostInitAuthPolicies for a description of the authorization policy
-		// for the v0 NV index.
-		if err := tpm.PolicyCommandCode(policySession, tpm2.CommandNVIncrement); err != nil {
+	for i := from; i < to; i++ {
+		// Compute a digest for signing with the update key - we have to do this
+		// on every loop iteration because the digest includes the session nonce
+		// which changes each time the session is used.
+		h := signDigest.NewHash()
+		h.Write(policySession.NonceTPM())
+		binary.Write(h, binary.BigEndian, int32(0)) // expiration
+
+		// Sign the digest
+		var signature tpm2.Signature
+		switch k := key.(type) {
+		case *rsa.PrivateKey:
+			sig, err := rsa.SignPSS(rand.Reader, k, signDigest.GetHash(), h.Sum(nil), &rsa.PSSOptions{SaltLength: rsa.PSSSaltLengthEqualsHash})
+			if err != nil {
+				return xerrors.Errorf("cannot sign authorization: %w", err)
+			}
+			signature = tpm2.Signature{
+				SigAlg: tpm2.SigSchemeAlgRSAPSS,
+				Signature: &tpm2.SignatureU{
+					RSAPSS: &tpm2.SignatureRSAPSS{
+						Hash: signDigest,
+						Sig:  tpm2.PublicKeyRSA(sig)}}}
+		case *ecdsa.PrivateKey:
+			sigR, sigS, err := ecdsa.Sign(rand.Reader, k, h.Sum(nil))
+			if err != nil {
+				return xerrors.Errorf("cannot sign authorization: %w", err)
+			}
+			signature = tpm2.Signature{
+				SigAlg: tpm2.SigSchemeAlgECDSA,
+				Signature: &tpm2.SignatureU{
+					ECDSA: &tpm2.SignatureECDSA{
+						Hash:       signDigest,
+						SignatureR: sigR.Bytes(),
+						SignatureS: sigS.Bytes()}}}
+		default:
+			panic("invalid private key type")
+		}
+
+		if cb != nil {
+			if err := cb(policySession); err != nil {
+				return err
+			}
+		}
+
+		if _, _, err := tpm.PolicySigned(keyLoaded, policySession, true, nil, nil, 0, &signature); err != nil {
 			return xerrors.Errorf("cannot execute assertion to increment counter: %w", err)
 		}
-		if err := tpm.PolicyNvWritten(policySession, true); err != nil {
+		if err := tpm.PolicyOR(policySession, c.authPolicies); err != nil {
```

Comment:
this PolicyOR is done once per for loop but always with the same value, I suppose it's needed this way but maybe it merits a comment